zebrad/src/components/inbound.rs

```diff
@@ -110,96 +131,131 @@ impl Service<zn::Request> for Inbound {
+        use oneshot::error::TryRecvError;
+
         // Check whether the network setup is finished, but don't wait for it to
         // become ready before reporting readiness. We expect to get it "soon",
         // and reporting unreadiness might cause unwanted load-shedding, since
         // the load-shed middleware is unable to distinguish being unready due
         // to load from being unready while waiting on setup.
-        if let Some(mut rx) = self.network_setup.take() {
-            use oneshot::error::TryRecvError;
-            match rx.try_recv() {
-                Ok((outbound, address_book)) => {
-                    self.outbound = Some(outbound.clone());
-                    self.address_book = Some(address_book);
-                    self.network_setup = None;
-                    self.downloads = Some(Box::pin(Downloads::new(
-                        Timeout::new(outbound, BLOCK_DOWNLOAD_TIMEOUT),
-                        Timeout::new(self.verifier.clone(), BLOCK_VERIFY_TIMEOUT),
-                        self.state.clone(),
-                    )));
-                }
-                Err(TryRecvError::Empty) => {
-                    self.network_setup = Some(rx);
+        if matches!(self.network, Setup::AwaitingNetwork { .. }) {
+            // Unfortunately, we can't match, swap, and destructure at the same time
+            let mut awaiting_state = Setup::FailedInit;
+            mem::swap(&mut self.network, &mut awaiting_state);
+            if let Setup::AwaitingNetwork {
+                mut network_setup,
+                verifier,
+            } = awaiting_state
+            {
+                match network_setup.try_recv() {
+                    Ok((outbound, address_book)) => {
+                        let downloads = Box::pin(Downloads::new(
+                            Timeout::new(outbound, BLOCK_DOWNLOAD_TIMEOUT),
+                            Timeout::new(verifier, BLOCK_VERIFY_TIMEOUT),
+                            self.state.clone(),
+                        ));
+                        self.network = Setup::Initialized {
+                            address_book,
+                            downloads,
+                        };
+                    }
+                    Err(TryRecvError::Empty) => {
+                        // There's no setup data yet, so keep waiting for it
+                        self.network = Setup::AwaitingNetwork {
+                            network_setup,
+                            verifier,
+                        };
+                    }
+                    Err(error @ TryRecvError::Closed) => {
+                        // Mark the service as failed, because network setup failed
+                        error!(?error, "inbound network setup failed");
+                        self.network = Setup::FailedRecv {
+                            failed_setup: network_setup,
+                        };
+                        return Poll::Ready(Err(error.into()));
```

Comment:
Where does this returned error go? And more importantly, does the receiver of this error also log the error? My preference would be to have errors be logged in the most central location we can, once it's been bubbled up the stack as high as it reasonably can. Otherwise it gets difficult to make sure we're logging each error once and exactly once.

Reply:
Inbound service readiness errors get turned into a `PeerError::Overloaded` error by the `Connection`.
The original error is dropped without being logged.

```rust
        if self.svc.ready_and().await.is_err() {
            // Treat all service readiness errors as Overloaded
            self.fail_with(PeerError::Overloaded);
            return;
        }
```

Reply:
that doesn't seem good :/

Reply:
I opened a ticket to treat `TryRecvError::Closed` in `Inbound::poll_ready` as a fatal error (#1655).

Making that change doesn't help solve Zebra hangs, and the current logs are sufficient to identify these errors - if they ever happen. (I suspect they don't.)

So I don't think we need to cleanup the error handling in this PR (or in the current sprint).